### DISCUSSION TITLE
Copy-paste CadDocument loses data

### DISCUSSION THREAD

**Discussion by JorenPelfrene** [DomCR/ACadSharp#479] [Q&A]
Hi,
First of all: cheers for this library and your great work.

I noticed that some data gets lost if I 'copy-paste' a CadDocument using ACadSharp. E.g. a test file of 22kb ends up at 6kb when I perform a simple read-write operation:

```
static void CopyDwg(string file, string outfile)
{
    CadDocument doc = DwgReader.Read(file);

    using (DwgWriter writer = new DwgWriter(outfile, doc))
    {
        writer.Write();
    }
}
```

It becomes an issue when I try to open the output dwg file in some programs. Autodesk Inventor shows a blank page for any dwg file written by ACadSharp and also in DWG Trueview some entities are not visible (even though they are in the file when examining again with ACadSharp).
So my question is what data is lost here, or what needs to be done to enable compatibility with different software?

[test.zip](https://github.com/user-attachments/files/17495381/test.zip)

Thanks!

**Comment by DomCR** [comment]
Hi @JorenPelfrene 

The document seems to be quite old, I would recomend to update the version by:

```C#
CadDocument.Header.Version = ACadVersion.AC1032;
```

About the losing of size/data, at the current state may be caused because some of the infrastructure like XRecords and dictionary entries like Materials or VisualStyles are not implemented and are not written in the final file, but if you are not using them it should not be a problem.

Hope this helps!

**Reply by JorenPelfrene** [reply]
Thanks for your response. I do have the same issue with newer files too though (AC1032). Most of them open up fine with DWG TrueView, but never with Inventor. There it's just a blank page and I wonder why. Perhaps I should ask in the Inventor community.

Best.

**Reply by DomCR** [reply]
Does Inventor provides any kind of log about the issue?

**Comment by JorenPelfrene** [comment]
What I noticed so far:
A file that is a straight copy with ACadSharp as in the code snippet before is not opened properly by Inventor. Somehow both the model and paper space are not recognized upon import.

But when you transfer the entities to a new CadDocument (as below) it can be opened. The funny thing is that it opens up in paper space by default, but it is possible to activate the model space. This is actually also the case with DWG TrueView.
AutoCAD .dwg files open in model space straight away. Is there a setting to 'activate' the model space for the document when it opens?

```
static void TransferDwg(string file, string outfile)
{
    CadDocument doc = DwgReader.Read(file);
    CadDocument transfer = new CadDocument();
    transfer.Header.Version = doc.Header.Version;

    List<Entity> entities = new List<Entity>(doc.Entities);
    foreach (var item in entities)
    {
        Entity e = doc.Entities.Remove(item);
        transfer.Entities.Add(e);
    }

    using (DwgWriter writer = new DwgWriter(outfile, transfer))
    {
        writer.Write();
    }
}
```

In any case, this method allows me to edit the dwg the way I need to do. It would be handy if it opens in model space straight away.

**Reply by JorenPelfrene** [reply]
And it does with  `transfer.Header.ShowModelSpace = true;`

**Reply by DomCR** [reply]
For what you say the error may be related the infrastructure of the document, dictionary entries, xrecords and so, these elements are implemented but still causing some issues due some relationships between the different objects.

Sorry but there is no quick fix unless you can provide more information like a log from Inventor when you open the file.

**Reply by JorenPelfrene** [reply]
Can't find a log so far. But the deep copy of entities to a new CadDocument works fine and is actually a cleaner way of working for me. However, if I find a solution for the first issue, I'll document it here.

Thanks for the help :)
